### PR TITLE
Attempt to fix USBCRM issue; reduce code size

### PR DESCRIPTION
**Pull request by mtc-mlx**
The changes here reduce the code size by 68 bytes (using GCC 12). Next I will see if I can squeeze MSOS 1.0 or 2.0 capability into the remaining 88 bytes.

Also, while trying to reduce the code size, I noticed the DFLL step size limits were set to zero, so I set them to 1, which is what MPLAB Harmony does when USBCRM is enabled. I don't have a Sparkfun board to test with, but I suspect this might be the reason it was unreliable—perhaps in the other cases, it was just a lucky batch of SAMD11s that had better tolerances on the DFLL in open-loop mode.

The other potentially-functional change is to the USB strobe registers. Accessing these with a bitfield seems incorrect, because the read-modify-write will actually just clear the entire register in the case of INTFLAG and EPSTATUSCLR.

I tested these changes on a SAMD11D14A-SSUT, including flashing the VCP demo application.

### CODE REVIEW

**Comment by majbthrd** [comment]
Thank you!  I appreciate how well-reasoned the changes are.

Regarding CRM mode and my cautionary note in README.md, I must confess that I seem to have misplaced the problem Sparkfun board; I was searching in vain for it this morning.  There are a non-trivial number of hardware units of assorted designs deployed using the existing bootloader, so I'd like to think it is more than a lucky batch.  However, perhaps the existing DFLL settings are unusable on an unlucky batch?

I'd like to take a little more time to test things (and hopefully find the missing PCB to confirm the alternate DFLL settings), but my initial tests are all positive.  I wanted to get back to you now so that you didn't think I was ignoring the PR.

In the interim, when I went to test the code, I found that I had uncommitted changes from a year ago.  The intention had been to refactor the code to make it more flexible to pick a different app origin, but a pleasant side-effect was to save four bytes.  Four is a lot less than your simplifications have netted, but in case that small change helps you reach your goals, I've dusted it off and uploaded it as PR #6

**Comment by mtc-mlx** [comment]
Thanks for your feedback, and no hurry… neat idea to return the application address in a register btw.

I also tried to make the DFU request handling a bit more robust, but unfortunately could not make it fit in 1k. Even with some further shortcuts in the HW initialization it’s still 136 bytes too big. :( So those changes are not yet committed. For example, the DFU spec doesn’t require the host to send exactly 64 bytes per transfer, and technically the bootloader is supposed to wait for a USB reset after the last get-status request, before restarting into the application.

While I was at it, I added checks on the NVMCTL error interrupt, although tbh I’m not sure the hardware can even detect a “bad row”.

All that would not really justify the extra code space for some minor non-user-facing improvements, but I estimate that adding MSOS and string descriptor support would require increasing the bootloader size anyway (perhaps to 1.25k or 1.5k instead of 1k). And MSOS descriptors are a huge improvement on the (Windows) user’s end. Would you be open to increasing the reserved size to add these features—possibly with a note to use the older version if the user really needs the extra 512 bytes of space?

One other note about flash—in principle I think the bootloader should really be executing from SRAM. If the CPU attempts to fetch instructions from flash during an erase or write, it will stall the AHB, including SRAM access by the USB DMA master. I think the current code works because the stall only happens when BK1RDY=0, so the USB doesn’t try to do any DMA operations. Fortunately this is an easy fix (just put everything in ramfunc).

**Comment by majbthrd** [comment]
With the DFU specification in one hand and dfu-util source code in the other, I arrived at the present implementation using [Muntzing](https://en.wikipedia.org/wiki/Muntzing); there's no denying that this DFU implementation does deliberately cut corners to make it fit.

Regarding would-be 1.25k or 1.5k bootloaders, the ARM v6-M VTOR (Vector Table Offset Register) does allow multiples of 128 bytes.  However, the SAMDx BOOTPROT only provides for power-of-two sizes, so choosing such a size would mean forgoing write-protection.

I realize that it is tilting at windmills, but it really irks me that Microsoft's claimed objective of MSOS is "to accommodate manufacturers whose devices do not fit into the current set of USB device classes".  This is laughable IMHO, since DFU *IS* a core USB device class (with the author list including Microsoft personnel) but Windows acts like it is foreign. :(

Because Microsoft alone has contrived to make DFU painful, if larger bootloaders are feasible for your application, my recommendation would be to go HID instead.  My objective with this project was to be within 1024 bytes.  However, the great thing about open source is that you can always try your own approach.

Regarding ramfunc, and at the risk of stating the obvious: a symbol can be declared in any section name the programmer desires, but without corresponding code in startup to copy that re-located code, it won't be there.  The code size reduction included simplified startup code, so there would be a code size increase to accommodate the copy to RAM.

**Comment by mtc-mlx** [comment]
> Regarding would-be 1.25k or 1.5k bootloaders, the ARM v6-M VTOR (Vector Table Offset Register) does allow multiples of 128 bytes. However, the SAMDx BOOTPROT only provides for power-of-two sizes, so choosing such a size would mean forgoing write-protection.

Ah, good point. I'll see what I can do to compact it although I'm not too hopeful. I really want to avoid the "unable to read DFU status after completion" message from dfu-util though.

> I realize that it is tilting at windmills, but it really irks me that Microsoft's claimed objective of MSOS is "to accommodate manufacturers whose devices do not fit into the current set of USB device classes". This is laughable IMHO, since DFU IS a core USB device class (with the author list including Microsoft personnel) but Windows acts like it is foreign. :(

Indeed, ~the~ Microsoft's logic here is a bit flawed...

> Because Microsoft alone has contrived to make DFU painful,

But that's just not true. For one, it's just as difficult for an end-user (without sudo privileges) to use DFU on Linux because AFAIK no one ships udev rules for it by default, and probably won't do so in the name of security. In my case, I can probably get an account under my employer on the LVFS and distribute firmware thanks to fwupd, but if I were a hobbyist without a $6000 USB vendor ID, I'm not sure that would be possible.

I also doubt it was deliberate malice on Microsoft's part; more likely the employees who contributed to the DFU spec were not responsible for creating a Windows userspace API, and by the time someone realized it would be a good idea to have such a thing, it was too late and manufacturers were already shipping their own kernel-mode drivers for DFU.

In fact, I think it makes a lot of sense for a device to be able to indicate in its descriptors that its interface(s) can safely be accessed by userspace applications. And this is by nature platform-specific, so MSOS 2.0 really doesn't seem so bad to me (except that the descriptors are rather large).

> if larger bootloaders are feasible for your application, my recommendation would be to go HID instead. My objective with this project was to be within 1024 bytes. However, the great thing about open source is that you can always try your own approach.

Indeed, despite its ambiguities and flaws, I would prefer to stick with the DFU specification rather than HID.

> Regarding ramfunc, and at the risk of stating the obvious: a symbol can be declared in any section name the programmer desires, but without corresponding code in startup to copy that re-located code, it won't be there. The code size reduction included simplified startup code, so there would be a code size increase to accommodate the copy to RAM.

If I'm not mistaken, ramfunc is included in .data, so there is no extra code required to perform the copy, right?

**Comment by majbthrd** [comment]
> > Regarding ramfunc, and at the risk of stating the obvious: a symbol can be declared in any section name the programmer desires, but without corresponding code in startup to copy that re-located code, it won't be there. The code size reduction included simplified startup code, so there would be a code size increase to accommodate the copy to RAM.
> 
> If I'm not mistaken, ramfunc is included in .data, so there is no extra code required to perform the copy, right?

Quite right.  I stand corrected.

**Comment by mtc-mlx** [comment]
OK, I _just_ managed to squeeze my changes into 1k. I committed my work-in-progress overhaul to the mtc/dfu_robustness branch in my fork. Some of the extra DFU protocol checks are probably overkill. I still need to find a way to flush the page buffer safely in response to a zero-length download request (to avoid sending the write command before the address register is initialized). The four bytes saved by your PR #6 will probably be help with that. :)

I also tried using ramfunc; for some reason it changed the total size rather significantly (not yet sure why).

After all that, I learned that WinUSB does not support USB reset signaling, so I need to use something else to trigger the reset... :(

**Comment by mtc-mlx** [comment]
I separated the non-invasive optimizations and added them to this PR. The double-tap optimization isn't tested yet.

**Comment by mtc-mlx** [comment]
The changes in this PR have been tested successfully on my board so I think it's ready for your review.

I have additional changes queuing up in my mtc/dfu_robustness_v2 branch, but they are perhaps more controversial, so I'll open a separate PR once this one is merged.

**Comment by mtc-mlx** [comment]
I realized I had made a couple mistakes in the linker script: I renamed the section to "bl_info" in bootloader.c but forgot to update the input section in the linker script. I also realized that using the location counter to set the bl_info location was much more kludgy than just declaring a separate memory region, so that's fixed now too.

**Comment by majbthrd** [comment]
You are making changes faster than I have the time to adequately review them.  There are some neat optimization tricks in there, but I'm also very much NOT OK with stuff like the aptly-named "USE_RISKY_SHORTCUT".  Let's put these ongoing experiments in a newly-created branch called "experimentation" for the time being.  Thanks!

**Comment by mtc-mlx** [comment]
Please don't feel pressured to accept changes you don't want! I just wanted to share what I can. Indeed, some extra features I probably won't be able to publish (at least not right now).

There are probably less risky ways to implement the "risky shortcut" by using a second linking step, but probably not worth the 4 bytes.

**Comment by mtc-mlx** [comment]
Btw, returning the app origin from bootloader() actually increases the code size for me with GCC (I assume that's not the case with Clang?). Using a variable probably forces extra load-store instructions whereas the literal can be expressed with a MOVS followed by a shift.

File: startup.s [hunk 1]
```diff
@@ -51,31 +52,46 @@ _vectors:
...
   bcc zero_loop
 
+  /* Only r4-r8 saved by callee */
   bl bootloader
 
   /* if bootloader returns, we proceed to the user app (origin address now in r0) */
 
-  ldr r1, =0xE000ED08 /* VTOR register */
-  str r0, [r1] /* point VTOR to user app */
+  str r5, [r4] /* point VTOR to user app */
```
Comment (mtc-mlx) [inline startup.s:94]: This doesn't look right to me

**Comment by mtc-mlx** [comment]
@majbthrd I'm going to revert the changes from #6 in my branch since they increase the code size with GCC. I will try a different approach for the same objective.

File: startup.s (hunk 1 above)
Comment (mtc-mlx) [reply startup.s:94]: Indeed, the VTOR is zero which will break any application that uses interrupts. @majbthrd please give me a chance to review the changes to my PR before merging it next time.

**Pull request merged**